### ISSUE TITLE
Allow to load C functions on a read-only replica

### ISSUE THREAD

**Issue by rtokarev** [tarantool/tarantool#4642]
Currently, it's not possible to load C function in a read-only instance. Because `box.schema.func.create` method needs to write to a persistent system space `_func`.

Such a limitation, for example, disallows to use different sets of C functions on master and replicas. But on the other hand, LUA functions don't have such a limitation.

**Comment by cyrillos** [comment]
1) Could you please point me a scenario where a different set of C functions is **really** needed on ro box?
2) Could you please elaborate what you mean by LUA functions. Do you mean ones you create in console of RO box?

Currently read only box behaves as it should -- any attempt to modify DB content is refused because we're in RO mode. And I really don't like the idea of changing it. There must be a very strong reason for that.

**Comment by Totktonada** [comment]
A bit more context:

* A Lua function may be called when it is in `_G` (global Lua namespace) and appropriate permissions are granted.
* A Lua function may be registered in the `_func` system space, which requires an instance to be in the read-write mode.
* In recent tarantool versions a Lua function may also be persisted within _func and so replicated.

For a C function (not ffi, not Lua/C, but one that is [written to serve the iproto call request](https://www.tarantool.io/en/doc/latest/tutorials/c_tutorial/)) the only way to make it callable via iproto is to register the function in _func. Which, in turn, requires the instance be read-write.

I don't in touch this the code here, but I would look whether we can give a flag like `persist_def = false` to `box.schema.func.create()` to omit creating of _func record and hold the function information only in runtime.

Alternative: Postpone writting to _func when an instance is in read-only mode (until it'll be in read-write).

Alternative: Add a new space like _func, but a temporary one. Add lookup over this space if a function is not found in _func.

The latter looks best for me, but I don't know how self-restricting we should be in adding new system spaces.

**Comment by Gerold103** [comment]
I think it must be a more general thing. The request seems to be about just loading C functions anyhow. And I would look into doing it as close to plain Lua functions as possible (plan - not registered anywhere, just defined a script). Something like `require('cfunc').load(path_to_so_file, function_name)`. Then it would return a callable object which behaves like a function. So as it wouldn't be different from doing the same in Lua, where you can carry function pointer like a variable, like an object.

However all depends on what @rtokarev meant by his request.

**Comment by cyrillos** [comment]
Yes. I need a precise description of what to be implemented. I can start implementing whta you've just pointed but i fear of wasting time if @rtokarev meant something different. Thus waiting for a feedback first.

**Comment by rtokarev** [comment]
Our application consists of few functions, some implemented using LUA API and some using C API with calls into LUA. Initially, all functions were implemented in LUA but due to some reasons, we rewrited some to C.

Imagine, we implement a new C function. To make it callable through TNT_OP_CALL we need to add it using box.schema.func.add. And here is a problem, it's not possible to add it to a read-only tarantool, but on the other hand we are able to add new LUA functions using eval or restart.

So, now we update our application on a master and then on replicas.

When starting this issue I thought that this different behavior is a problem, because tarantool allows us to update LUA functions regardless of read-only status, but not allow C ones. Today I think that it's a good practice to update the application on a master first and then on replicas, because if something goes wrong with a new version, we always have a good replica to switch to.

So it's your choice whether or not to change such behavior.

In my opinion, tarantool should provides the same functionality for LUA and C API. Maybe it should be another way, except box.schema.func.add, to load C functions in the same manner as LUA ones.
